### PR TITLE
Generate: deprecate generation relying on default `max_length`

### LINKED ISSUE #17868: Calling `generate` on a `T5ForConditionalGeneration` returns `n` tokens but `n-1` scores
### System Info

```shell
- `transformers` version: 4.20.1
- Platform: Linux-5.4.0-113-generic-x86_64-with-glibc2.17
- Python version: 3.8.13
- Huggingface_hub version: 0.8.1
- PyTorch version (GPU?): 1.11.0+cu102 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no
```

### Who can help?

@patrickvonplaten, @Narsil

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer
import torch

if __name__ == '__main__':
    torch.manual_seed(0)
    tokenizer = AutoTokenizer.from_pretrained('t5-small')
    model = AutoModelForSeq2SeqLM.from_pretrained('t5-small')

    input = tokenizer.encode("I enjoy walking with my cute dog", return_tensors='pt')
    result = model.generate(
        input,
        max_new_tokens=15,
        do_sample=True,
        return_dict_in_generate=True,
        output_scores=True,
    )

    print(len(result["scores"]))
    for sequence in result["sequences"]:
        print(len(sequence))
        print(tokenizer.decode(sequence))
```

Output:
```
15
16
<pad> Ich, liebe es, mes lustig beim laufen
```

### Expected behavior

I would have expected to have up to 15 tokens (as `max_new_tokens=15`) and `len(result["scores"]) == len(result["sequences"][0])`. However, the size of the returned sequence of tokens is always `len(result["scores"]) + 1`. In addition, if `max_new_tokens` is reached we have `len(result["sequences"][0]) == max_new_tokens + 1`.

When looking at the decoded sequence, there is always a pad token at the beginning.

I don't know if this is necessarily a bug but this behaviour is somewhat confusing, especially when trying to compute the probability of the sequence given scores.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

(EDITED)

This PR applies the outcome of two discussions:
1. [Favor the use of `max_new_tokens`, as opposed to `max_length`](https://github.com/huggingface/transformers/issues/17414#issuecomment-1148836312). This introduces the `max_new_tokens` to TF and FLAX.
2. [Update the docstring about `scores`](https://github.com/huggingface/transformers/issues/17868#issuecomment-1171148445), to be clear that its length depends on `max_new_tokens` and other factors.

Review suggestion: review PT first. TF and Flax are mostly copy/paste.

Closes #17868

### CODE REVIEW

File: src/transformers/generation_flax_utils.py [hunk 1]
```diff
@@ -209,8 +211,10 @@ def generate(
 
             input_ids (`jnp.ndarray` of shape `(batch_size, sequence_length)`):
                 The sequence used as a prompt for the generation.
-            max_length (`int`, *optional*, defaults to 20):
-                The maximum length of the sequence to be generated.
+            max_length (`int`, *optional*, defaults to `model.config.max_length`):
+                **DEPRECATED** The maximum length of the sequence to be generated. Prefer the use of `max_new_tokens`.
```
Comment (sgugger) [inline src/transformers/generation_flax_utils.py:215]:
```suggestion
                **DEPRECATED** The maximum length of the sequence to be generated. Prefer the use of `max_new_tokens` which ignore the number of tokens in the prompt.
```
(Here and in all other docstrings.)

File: src/transformers/generation_flax_utils.py [hunk 2]
```diff
@@ -283,6 +280,41 @@ def generate(
...
+        # if `max_new_tokens` is passed, but not `max_length` -> set `max_length = max_new_tokens`
+        input_ids_seq_length = input_ids.shape[-1]
+        if max_length is not None:
+            warnings.warn(
+                "The `max_length` argument is deprecated and will be removed in v5. Use `max_new_tokens` instead.",
+                FutureWarning,
+            )
+        if max_length is None and max_new_tokens is not None:
+            max_length = max_new_tokens + input_ids_seq_length
+        elif max_length is not None and max_new_tokens is not None:
```
Comment (sgugger) [inline src/transformers/generation_flax_utils.py:293]: It is a bit weird to me that the code uses the deprecated argument. The code should use `max_new_tokens` if possible, and then we set `max_new_tokens` when `max_length` is provided, along with issuing the deprecation warning.

File: src/transformers/generation_flax_utils.py [hunk 3]
```diff
@@ -283,6 +280,41 @@ def generate(
...
+            # Both are set, this is odd, raise a warning
+            warnings.warn(
+                "Both `max_length` and `max_new_tokens` have been set "
+                f"but they serve the same purpose. `max_length` {max_length} "
+                f"will take priority over `max_new_tokens` {max_new_tokens}.",
+                UserWarning,
+            )
```
Comment (sgugger) [inline src/transformers/generation_flax_utils.py:300]: This should be a `ValueError` IMO.

File: src/transformers/generation_flax_utils.py [hunk 4]
```diff
@@ -283,6 +280,41 @@ def generate(
...
+            raise ValueError(
+                f"Unfeasable length constraints: the minimum length ({min_length}) is larger than the maximum "
+                f"length ({max_length})"
+            )
+        if input_ids_seq_length >= max_length:
+            input_ids_string = "decoder_input_ids" if self.config.is_encoder_decoder else "input_ids"
+            logger.warning(
+                f"Input length of {input_ids_string} is {input_ids_seq_length}, but ``max_length`` is set to"
+                f" {max_length}. This can lead to unexpected behavior. You should consider increasing"
+                "``max_new_tokens``."
```
Comment (sgugger) [inline src/transformers/generation_flax_utils.py:315]: No need for double backtiks here.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this. Made comments on the Flax file which apply to the others. I'll let @patrickvonplaten comment on whether we should remove this `max_length` argument in v5 or not, but `max_new_tokens` makes way more sense to me!

**Comment by gante** [comment]
@patrickvonplaten @sgugger I've implemented the changes from the comments above 👍 Related to the `max_length`/`max_new_tokens`, the behavior is now the following:
- Neither are set ➡️  raise a deprecation warning (will be an exception in v5), nudge towards `max_new_tokens`, and use the default for `max_length`;
- Both are set ➡️ raise an exception;
- `max_length` is set ➡️ nothing happens
- `max_new_tokens` is set ➡️ updates `max_length` according to the prompt length

File: src/transformers/generation_flax_utils.py [hunk 5]
```diff
@@ -209,8 +211,11 @@ def generate(
 
             input_ids (`jnp.ndarray` of shape `(batch_size, sequence_length)`):
                 The sequence used as a prompt for the generation.
-            max_length (`int`, *optional*, defaults to 20):
-                The maximum length of the sequence to be generated.
+            max_length (`int`, *optional*, defaults to `model.config.max_length`):
+                The maximum length of the sequence to be generated. Prefer the use of `max_new_tokens`, which ignores
+                the number of tokens in the prompt.
+            max_new_tokens (`int`, *optional*, defaults to None):
```
Comment (sgugger) [inline src/transformers/generation_flax_utils.py:217]:
```suggestion
            max_new_tokens (`int`, *optional*):
```
We don't indicate it as it's implied by "optional".

File: src/transformers/generation_flax_utils.py [hunk 6]
```diff
@@ -283,6 +281,40 @@ def generate(
...
+                "removed in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length"
+                "of the generation.",
+                UserWarning,
+            )
+        elif max_length is None and max_new_tokens is not None:
+            max_length = max_new_tokens + input_ids_seq_length
+        elif max_length is not None and max_new_tokens is not None:
+            raise ValueError(
+                "Both `max_new_tokens` and `max_length` have been set but they serve the same purpose -- setting a "
+                "limit to the generated output length. Please refer to the documentation for more information."
```
Comment (sgugger) [inline src/transformers/generation_flax_utils.py:299]:
```suggestion
                "limit to the generated output length. Remove one of those arguments. Please refer to the documentation for more information."
```

**Review by sgugger — APPROVED** [review]
Made some comments that should be applied on all three frameworks.

LGTM. thanks for iterating on this!

**Comment by gante** [comment]
@Narsil that seems like an interesting use case 🤔 What would be the default arguments, if setting neither was a valid option? Perhaps a Stopping condition that would target this use case?

From what I've seen, the default arguments are one of the biggest pain points for new users (and even some experienced users), so whatever decisions we make in this PR should also cover that!

File: src/transformers/generation_flax_utils.py [hunk 7]
```diff
@@ -209,8 +211,11 @@ def generate(
 
             input_ids (`jnp.ndarray` of shape `(batch_size, sequence_length)`):
                 The sequence used as a prompt for the generation.
-            max_length (`int`, *optional*, defaults to 20):
-                The maximum length of the sequence to be generated.
+            max_length (`int`, *optional*, defaults to `model.config.max_length`):
+                The maximum length of the sequence to be generated. Prefer the use of `max_new_tokens`, which ignores
+                the number of tokens in the prompt.
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:216]: Note that this is also the case for `max_length` if the model is an encoder-decoder (that the prompt length is ignored). For encoder-decoder models `max_length = max_new_tokens + 1` holds almost always true. Maybe
```suggestion
                 The maximum length the generated tokens can have corresponds to the length of the input prompt + `max_new_tokens`. In general, prefer the use of `max_new_tokens`, which ignores
```

File: src/transformers/generation_flax_utils.py [hunk 8]
```diff
@@ -209,8 +211,11 @@ def generate(
...
+            max_new_tokens (`int`, *optional*):
+                The maximum numbers of tokens to generate, ignoring the number of tokens in the prompt.
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:218]: Nice  & precise doc string

File: src/transformers/generation_flax_utils.py [hunk 9]
```diff
@@ -283,6 +281,41 @@ def generate(
             # prepare decoder_input_ids for generation
             input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
 
+        # Prepare `max_length` depending on other stopping criteria.
+        input_ids_seq_length = input_ids.shape[-1]
+        if max_length is None and max_new_tokens is None:
+            warnings.warn(
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:287]: Very nice! I like it. It'll create a lot of warnings in the beginning, but there is no way around it

File: src/transformers/generation_flax_utils.py [hunk 10]
```diff
@@ -283,6 +281,41 @@ def generate(
...
+                "Neither `max_length` nor `max_new_tokens` have been set, `max_length` will default to "
+                f"{self.config.max_length} (`self.config.max_length`). This behavior is deprecated and will be "
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:289]:
```suggestion
                f"{self.config.max_length} (`self.config.max_length`). Controlling `max_length` via the config is deprecated and `max_length` will be "
```

File: src/transformers/generation_flax_utils.py [hunk 11]
```diff
@@ -283,6 +281,41 @@ def generate(
...
+                "removed in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length "
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:290]:
```suggestion
                "removed from the config in v5 of Transformers -- we recommend using `max_new_tokens` to control the maximum length "
```

File: src/transformers/generation_flax_utils.py [hunk 12]
```diff
@@ -283,6 +281,41 @@ def generate(
...
+                "of the generation.",
+                UserWarning,
+            )
+        elif max_length is None and max_new_tokens is not None:
+            max_length = max_new_tokens + input_ids_seq_length
+        elif max_length is not None and max_new_tokens is not None:
+            raise ValueError(
+                "Both `max_new_tokens` and `max_length` have been set but they serve the same purpose -- setting a"
+                " limit to the generated output length. Remove one of those arguments. Please refer to the"
+                " documentation for more information."
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:300]: (nit) What documentation? Can we provide a link here?

File: src/transformers/generation_tf_utils.py [hunk 13]
```diff
@@ -1561,21 +1563,48 @@ def _generate(
                 model_kwargs=model_kwargs,
             )
 
-        if input_ids.shape[-1] >= max_length:
+        # 5. Prepare `max_length` depending on other stopping criteria.
+        input_ids_seq_length = input_ids.shape[-1]
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1567]: same comments as for Flax

File: src/transformers/generation_tf_utils.py [hunk 14]
```diff
@@ -1332,8 +1339,11 @@ def _generate(
             input_ids (`tf.Tensor` of `dtype=tf.int32` and shape `(batch_size, sequence_length)`, *optional*):
                 The sequence used as a prompt for the generation. If `None` the method initializes it with
                 `bos_token_id` and a batch size of 1.
-            max_length (`int`, *optional*, defaults to 20):
-                The maximum length of the sequence to be generated.
+            max_length (`int`, *optional*, defaults to `model.config.max_length`):
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:1342]: nice change

**Review by patrickvonplaten — APPROVED** [review]
Very nice work @gante ! Thanks a lot for transforming the discussion into an actionable PR! Everything looks good to me. Some nitpicks regarding the wording, but apart from this it looks all good!

**Pull request merged**